Proposals/linear-haskell-proposal-faq.org

```diff
@@ -0,0 +1,213 @@
+
+* Why does this proposal need to change Core?
+
+There are two reasons:
+- In GHC, the type ~Type~ is shared between the surface language and
+  Core. So modifying the types implies that Core is modified as well.
+- More importantly, because Core is a good check that our
+  implementation works as intended. That is, a linearly typed Core
+  will ensure that linearly-typed programs are indeed desugared to
+  linearly-typed programs in Core. And that optimisations do not
+  destroy linearity.
+
+* Isn't a linear function that diverges unsound?
+
+Our proposed type system allows you to give a linear type to list
+concatenation:
+#+BEGIN_SRC haskell
+  (++) :: [a] ->. [a] ->. [a]
+#+END_SRC
+
+So we can write
+#+BEGIN_SRC haskell
+  strange :: Int ->. [Int]
+  strange x = (repeat 1) ++ [x]
+#+END_SRC
+This sounds like it's breaking the promise made by ~strange~ that it
+will consume its argument exactly once. Since ~x~ will /never/ be
+consumed!
+
+But there is no cause for panic: ~strange~ only promises to consume
+~x~ if its result is consumed. Which will never happen since the
+result of ~strange~ is infinite. The properties of linear types which
+we need are not compromised by non-termination.
+
+* In the motivations of the proposal, there is a linear IO monad. Isn't linear IO unsound in presence of exception?
+
+It is not, but as always, we need to careful about how we type
+primitives of the ~IO~ monad. For example, [[https://github.com/tweag/linear-base/blob/007b884ebb0e3182ea73e450683f9660b7a92f40/src/System/IO/Linear.hs#L144-L146][~catch~]] should not have
+a linear type. It should be typed as follows:
+#+BEGIN_SRC haskell
+  catch
+    :: Exception e
+    => IO (Unrestricted a) -> (e -> IO (Unrestricted a)) -> IO (Unrestricted a)
+#+END_SRC
+It means that neither the body nor the handler of ~catch~ can capture
+linear variables. Otherwise, you could lose track of them.
+
+If, say, the body of ~catch~ was linear, we could write:
+#+BEGIN_SRC haskell
+  oops :: a ->. IO (Unrestricted ())
+  oops x =
+    catch
+      (throwIO "This is bad" >> oops a)
+      (\_ -> return $ Unrestricted ())
+#+END_SRC
+
+* Don't linear guarantees degrade to affine in the presence of exceptions?
+
+When an exception is raised during the consumption of ~f u~, ~u~ may
+not have been fully consumed. Functions whose argument is consumed /at
+most/ once when their result is consumed exactly once are called
+affine.
+
+Linear functions do have the property that: /if their result is
+consumed at most once, then their argument is consumed at most
+once/. This is a more relevant phenomenon in case exceptions are
+raised: exceptions interrupt the consumption of the result, which is
+only partial.
+
+In fact, linear functions have the property that if their result is
+consumed with multiplicity X, then their argument is also consumed
+with multiplicity X. Whatever choice of multiplicity we add to the
+language. Which can be internalised in the language as follows
```

Comment:
"Whatever choice of multiplicity we add to the language" -- I think it's a little unclear what this mean to the general audience.